iothub/device/src/Transport/Mqtt/OrderedTwoPhaseWorkQueue.cs

```diff
@@ -46,16 +46,22 @@ public Task CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
                 throw new IotHubException("Nothing to complete.", isTransient: false);
             }
 
-            IncompleteWorkItem incompleteWorkItem = _incompleteQueue.Peek();
-            if (incompleteWorkItem.Id.Equals(workId))
+            if (_incompleteQueue.TryDequeue(out IncompleteWorkItem incompleteWorkItem))
             {
-                _incompleteQueue.Dequeue();
-                return _completeWorkAsync(context, incompleteWorkItem.WorkItem);
-            }
+                if (incompleteWorkItem.Id.Equals(workId))
+                {
+                    return _completeWorkAsync(context, incompleteWorkItem.WorkItem);
+                }
 
-            throw new IotHubException(
```

Comment:
Q - is this exception caught higher in the stack and then the operation retried? If yes, then the new implementation will show a different behavior since previously the queue was unchanged but now the queue will have been modified (peek vs dequeue).

Reply:
No retry happens here, in case this exception is thrown, everything shuts down and the queue will be emptied out.